### PR TITLE
experimental wrappers for mmx/fp support

### PR DESCRIPTION
**Pull request by humeafo**
just for experiments and compiled, not functioning yet, no debugging is done.

### CODE REVIEW

File: include/cpu/i386/cpu.h [hunk 1]
```diff
@@ -113,22 +125,13 @@ typedef struct CPUX86State {
     uint16_t fpus;
```
Comment (vitalych) [inline include/cpu/i386/cpu.h:125]: Do you need to move these too? If not, could write a quick comment in the code

File: src/target-i386/cpu.h [hunk 2]
```diff
@@ -107,12 +107,67 @@ static inline void __WR_env_raw(CPUX86State *cpuState, unsigned offset, target_u
...
 #define WR_cpu(cpu, reg, value) __WR_env_raw(cpu, offsetof(CPUX86State, reg), (target_ulong) value, sizeof(cpu->reg))
+
+#define RR_cpu_fp80(cpu, reg) (__RR_env_floatx80(cpu, offsetof(CPUX86State, reg)))
+#define WR_cpu_fp80(cpu, reg, value) __WR_env_floatx80(cpu, offsetof(CPUX86State, reg), value)
+
+#define RR_cpu_dyn(p, size) ((__typeof__(*p)) __RR_cpu_dyn(p, size))
+#define WR_cpu_dyn(p, size, v) __WR_cpu_dyn(p, size, v)
+
+#if 1
+#define WR_reg(r, v)                    \
```
Comment (vitalych) [inline src/target-i386/cpu.h:149]: wrap the content of this macro in { .. }

File: src/target-i386/op_helper.c [hunk 3]
```diff
@@ -4267,20 +4276,20 @@ void helper_fprem1(void) {
         else
             q = (signed long long int) dblq;
 
-        env->fpus &= (~0x4700);             /* (C3,C2,C1,C0) <-- 0000 */
+        FPUS_W(FPUS & (~0x4700));             /* (C3,C2,C1,C0) <-- 0000 */
                                             /* (C0,C3,C1) <-- (q2,q1,q0) */
-        env->fpus |= (q & 0x4) << (8 - 2);  /* (C0) <-- q2 */
-        env->fpus |= (q & 0x2) << (14 - 1); /* (C3) <-- q1 */
-        env->fpus |= (q & 0x1) << (9 - 0);  /* (C1) <-- q0 */
+        FPUS_W(FPUS | (q & 0x4) << (8 - 2));  /* (C0) <-- q2 */
```
Comment (vitalych) [inline src/target-i386/op_helper.c:4281]: Make sure to put parenthesis around expressions, to make priority clear.

File: src/target-i386/op_helper.c [hunk 4]
```diff
@@ -5573,27 +5582,35 @@ static void update_sse_status(void) {
...
-    env->mxcsr = val;
+    MXCSR_W(val);
     update_sse_status();
 }
 
 void helper_enter_mmx(void) {
-    env->fpstt = 0;
+    FPSTT_W(0);
+#if defined(CONFIG_SYMBEX) && !defined(SYMBEX_LLVM_LIB)
+    WR_cpu(env, fptags, 0);
```
Comment (vitalych) [inline src/target-i386/op_helper.c:5599]: I would use macros here instead of having the #if defined()

File: src/target-i386/op_helper.c [hunk 5]
```diff
@@ -5966,3 +5983,43 @@ void helper_se_opcode(uint64_t opcode) {
 #endif
 }
 #endif
+
+#if defined(CONFIG_SYMBEX) && !defined(SYMBEX_LLVM_LIB)
+uint8_t RR_cpu_float_status(void* p, unsigned size) {
+    assert(size == 1 && "unexpected status field width");
+
+    if ((void*)env < p && (p - (void*)env) < offsetof(CPUArchState, eip)) {
```
Comment (vitalych) [inline src/target-i386/op_helper.c:5991]: Arithmetic on void pointers is undefined behavior.
https://stackoverflow.com/questions/3523145/pointer-arithmetic-for-void-pointer-in-c#3524270

Comment (vitalych) [reply src/target-i386/op_helper.c:5991]: I wonder why the compiler didn't catch it. Perhaps we need to add the undefined behavior sanitized flag to clang.

Comment (humeafo) [reply src/target-i386/op_helper.c:5991]: Yes, char* cast would be more portable, clang seems in order to keep compatibility of gcc, allow this.

File: src/target-i386/op_helper.c (hunk 4 above)
Comment (humeafo) [reply src/target-i386/op_helper.c:5599]: Yes, this is the result of fast prototying...

File: include/cpu/i386/cpu.h (hunk 1 above)
Comment (humeafo) [reply include/cpu/i386/cpu.h:125]: In the first version I moved them too, but later I moved them out to make debugging easier, maybe should moved too in the final version. The reason is I found fpuc was referenced in other files than the helpers file such as s2e-kvm-state.c, I'm not sure whether I should wrap them too or not.

File: src/target-i386/op_helper.c (hunk 3 above)
Comment (humeafo) [reply src/target-i386/op_helper.c:4281]: Sure.

**Pull request closed without merging**